sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala

```diff
@@ -0,0 +1,256 @@
+        ).resolveRelation(checkFilesExist = false))
+        .select("value").as[String](Encoders.STRING)
+    } else {
+      val charset = options.charset
+      val rdd = sparkSession.sparkContext
+        .hadoopFile[LongWritable, Text, TextInputFormat](paths.mkString(","))
+        .mapPartitions(_.map(pair => new String(pair._2.getBytes, 0, pair._2.getLength, charset)))
+      sparkSession.createDataset(rdd)(Encoders.STRING)
+    }
+  }
+}
+
+object WholeFileCSVDataSource extends CSVDataSource {
+  override val isSplitable: Boolean = false
+
+  override def readFile(
+      conf: Configuration,
+      file: PartitionedFile,
+      parser: UnivocityParser,
+      parsedOptions: CSVOptions): Iterator[InternalRow] = {
+    UnivocityParser.tokenizeStream(
+      createInputStream(conf, file.filePath),
+      parsedOptions.headerFlag,
+      parsedOptions.asParserSettings).flatMap { tokens =>
+        parser.convert(tokens)
+      }
+  }
+
+  override def infer(
+      sparkSession: SparkSession,
+      inputPaths: Seq[FileStatus],
+      parsedOptions: CSVOptions): Option[StructType] = {
+    val csv: RDD[PortableDataStream] = createBaseRdd(sparkSession, inputPaths, parsedOptions)
+    val firstRow = csv.mapPartitions { iter =>
+      // We need this header line always because even if `inferSchema` option is disabled,
+      // it needs to know the number of columns that are string types without schema inference.
+      iter.flatMap { lines =>
+        UnivocityParser.tokenizeStream(
+          createInputStream(lines.getConfiguration, lines.getPath()),
+          false,
+          parsedOptions.asParserSettings)
+      }
+    }.first()
+
+    val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
+    val header = makeSafeHeader(firstRow, caseSensitive, parsedOptions)
+    val tokenRDD = csv.mapPartitions { iter =>
+      iter.flatMap { lines =>
+        UnivocityParser.tokenizeStream(
+          createInputStream(lines.getConfiguration, lines.getPath()),
+          parsedOptions.headerFlag,
+          parsedOptions.asParserSettings)
+      }
+    }
+
+    Some(CSVInferSchema.infer(tokenRDD, header, parsedOptions))
+  }
+
+  private def createBaseRdd(
+      sparkSession: SparkSession,
+      inputPaths: Seq[FileStatus],
+      options: CSVOptions): RDD[PortableDataStream] = {
+    val paths = inputPaths.map(_.getPath)
+    val name = paths.mkString(",")
+    val job = Job.getInstance(sparkSession.sessionState.newHadoopConf())
+    FileInputFormat.setInputPaths(job, paths: _*)
+    val conf = job.getConfiguration
+
+    new BinaryFileRDD(
+      sparkSession.sparkContext,
+      classOf[StreamInputFormat],
+      classOf[String],
+      classOf[PortableDataStream],
+      conf,
+      sparkSession.sparkContext.defaultMinPartitions)
+      .setName(s"CSVFile: $name")
+      .values
+  }
+
+  private def createInputStream(config: Configuration, path: String): InputStream = {
```

Comment:
Might be nice to move this over to `CodecStreams` (with an appropriate name) and fix up other classes that use this pattern.